src/hooks/useSnackbar/useSnackbar.tsx

```diff
@@ -1,76 +1,98 @@
-import { snackbarTransitions } from '@/shared/components/Snackbar'
+import React, { createContext, useContext, useState, useRef, useEffect } from 'react'
+import { CSSTransition, TransitionGroup } from 'react-transition-group'
+import styled from '@emotion/styled'
+import faker from 'faker'
 import Snackbar from '@/shared/components/Snackbar/Snackbar'
-import { transitions } from '@/shared/theme'
-import { Global } from '@emotion/react'
-import React, { createContext, useContext, useEffect, useState } from 'react'
-import { CSSTransition } from 'react-transition-group'
+import { transitions, sizes } from '@/shared/theme'
 
 export type DisplaySnackbarArgs = {
   time?: number
-  variant?: 'success' | 'error' | 'info'
-  buttonText?: string
+  variant?: 'primary' | 'secondary'
+  icon?: 'success' | 'error' | 'info'
   message: string
+  subMessage?: string
+  actionText?: string
+  onActionClick?: () => void
 }
 
+type SnackbarsState = {
+  id: string
+} & Omit<DisplaySnackbarArgs, 'time'>
+
 type SnackbarContextValue = {
   displaySnackbar: (args: DisplaySnackbarArgs) => void
-  closeSnackbar: () => void
+  closeSnackbar: (id: string) => void
 }
 
 const SnackbarContext = createContext<SnackbarContextValue | undefined>(undefined)
 SnackbarContext.displayName = 'SnackbarContext'
 
+const SNACKBARS_LIMIT = 3
+
 export const SnackbarProvider: React.FC = ({ children }) => {
-  const [isVisible, setIsVisible] = useState(false)
-  const [snackbarOpts, setSnackbarOpts] = useState<DisplaySnackbarArgs | null>(null)
+  const [snackbars, setSnackbars] = useState<SnackbarsState[]>([])
+
+  // Snackbars state reference is needed for setTimeout to read
+  const snackbarsRef = useRef(snackbars)
+  snackbarsRef.current = snackbars
```

Comment:
Could you explain rationale behind it?

Reply:
@kdembler I had an issue with using the state directly in setTimeout, so had to put it in container - that solution is described in details here:
[https://github.com/facebook/react/issues/14010#issuecomment-433788147](https://github.com/facebook/react/issues/14010#issuecomment-433788147)

Reply:
Okay, so I may have been more explicit. I actually do understand the problem and I also think there's a better solution to it than this one. I was hoping you could try to explain the problem so we may arrive to the conclusion. Do you understand why this doesn't work with just using `snackbars` directly and why the ref helps us with this problem? I want to make sure you properly understand what happens before we get to the solution

Reply:
As I understand it, the problem is that setTimeout uses `snackbars` from the moment setTimeout is triggered, so the setState function inside operates on `snackbars` state that comes from the moment of triggering and not current state (after timeout). It is not a problem as long as nothing's changed after trigger (or state will change inside its scope) - but in our case it will happen. New snackbars will appear in between. Is that correct?

Reply:
Yep, sounds right! When we create a function that we pass as an argument to `setTimeout`, we pass `snackbars` reference to it. Since all of our data is immutable (when we do modification, we don't modify an existing array but rather replace the reference to it with a new one) this will not get updated if there's a change before the timeout executes.

Okay, so now that the problem is clear, let's get to the solution. When you have some setter from `useState`, you can pass a value to it (a new array), but you can also pass a function. This function will get executed with the current value of that state, and I guess it's available for purposes just like this one, when you need to update the state based on current value. We can use this and get rid of the ref:
```ts
setTimeout(() => {
  setSnackbars((currentSnackbars) => currentSnackbars.filter((snackbar) => snackbar.id !== id))
}, timeout)
```